File: templates/nodejs-api/src/routes/health.js
'use strict';

const { Router } = require('express');
const logger = require('../logger');

const router = Router();

let isReady = false;

// Signal readiness after warm-up (database connections, cache warming, etc.)
setTimeout(() => {
  isReady = true;
  logger.info('Service marked as ready');
}, parseInt(process.env.STARTUP_DELAY_MS || '2000', 10));

/**
 * Liveness probe — is the process alive?
 * Kubernetes restarts the pod if this returns non-200.
 */
router.get('/live', (req, res) => {
  res.status(200).json({
    status: 'alive',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
  });
});

/**
 * Readiness probe — is the service ready to handle traffic?
 * Kubernetes removes the pod from load balancer if this returns non-200.
 */
router.get('/ready', async (req, res) => {
  if (!isReady) {
    return res.status(503).json({
      status: 'not_ready',
      reason: 'Warming up',
      timestamp: new Date().toISOString(),
    });
  }

  // Check dependencies
  const checks = await runHealthChecks();
  const allHealthy = checks.every((c) => c.healthy);

  return res.status(allHealthy ? 200 : 503).json({
    status: allHealthy ? 'ready' : 'degraded',
    checks,
    timestamp: new Date().toISOString(),
  });
});

/**
 * Startup probe — has the application started?
 */
router.get('/startup', (req, res) => {
  res.status(isReady ? 200 : 503).json({
    status: isReady ? 'started' : 'starting',
    timestamp: new Date().toISOString(),
  });
});

async function runHealthChecks() {
  const checks = [];

  // Database check (if configured)
  if (process.env.DATABASE_URL) {
    checks.push(await checkDatabase());
  }

  // Redis check (if configured)
  if (process.env.REDIS_URL) {
    checks.push(await checkRedis());
  }

  return checks;
}

async function checkDatabase() {
  try {
    // Replace with actual DB health check
    return { name: 'database', healthy: true, latency_ms: 1 };
  } catch (err) {
    logger.warn({ err }, 'Database health check failed');
    return { name: 'database', healthy: false, error: err.message };
  }
}

async function checkRedis() {
  try {
    // Replace with actual Redis PING
    return { name: 'redis', healthy: true, latency_ms: 0 };
  } catch (err) {
    logger.warn({ err }, 'Redis health check failed');
    return { name: 'redis', healthy: false, error: err.message };
  }
}

module.exports = router;

File: templates/python-fastapi/app/health.py
"""Health check endpoints — required for Kubernetes probes."""

from __future__ import annotations

import time
from typing import Literal

from fastapi import APIRouter, status
from pydantic import BaseModel

health_router = APIRouter(prefix="/health", tags=["health"])

_startup_time = time.time()
_ready = False


class HealthResponse(BaseModel):
    status: Literal["alive", "ready", "degraded", "starting"]
    timestamp: str
    uptime_seconds: float | None = None
    checks: list[dict] | None = None


@health_router.get(
    "/live",
    response_model=HealthResponse,
    summary="Liveness probe",
    description="Returns 200 if the process is alive. Kubernetes restarts pod if this fails.",
)
async def liveness() -> HealthResponse:
    return HealthResponse(
        status="alive",
        timestamp=_now(),
        uptime_seconds=round(time.time() - _startup_time, 1),
    )


@health_router.get(
    "/ready",
    response_model=HealthResponse,
    summary="Readiness probe",
    description="Returns 200 if service is ready for traffic.",
)
async def readiness() -> HealthResponse:
    checks = await _run_checks()
    all_ok = all(c["healthy"] for c in checks)
    return HealthResponse(
        status="ready" if all_ok else "degraded",
        timestamp=_now(),
        checks=checks,
    )


@health_router.get("/startup", summary="Startup probe")
async def startup() -> dict:
    return {"status": "started", "timestamp": _now()}


async def _run_checks() -> list[dict]:
    checks = []
    # Add dependency checks here:
    # checks.append(await check_database())
    # checks.append(await check_redis())
    return checks


def _now() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()
